linux_os/guide/system/network/network-firewalld/ruleset_modifications/ensure_firewall_rules_for_open_ports/oval/shared.xml

```diff
@@ -1,76 +1,300 @@
+    <linux:state state_ref="state_firewalled_listening_inet_direct_tcp_ports" />
   </linux:inetlisteningservers_test>
 
-  <linux:inetlisteningservers_state id="state_firewalled_listening_inet_ports" version="1" comment="Checks listen ports has rule in firewalld">
-      <linux:local_port datatype="int" operation="equals"  var_ref="var_firewalled_ports" var_check="at least one"/>
-    </linux:inetlisteningservers_state>
+  <linux:inetlisteningservers_test id="test_listening_fw_inet_direct_ports_tcp_test" version="1"
+                                   check="at least one"
+                                   comment="Check all listening ports defined by service are firewalled">
+    <linux:object object_ref="obj_listening_inet_tcp_ports" />
+    <linux:state state_ref="state_firewalled_listening_inet_tcp_ports" />
+  </linux:inetlisteningservers_test>
+
+  <ind:variable_test id="test_var_firewalled_direct_tcp_ports_exists" check="all"
+                     check_existence="at_least_one_exists" version="1"
+                     comment="Check the existence of direct tcp port definitions.">
+    <ind:object object_ref="object_var_firewalled_direct_tcp_ports"/>
+  </ind:variable_test>
+
+  <ind:variable_test id="test_var_firewalled_tcp_ports_exists" check="all"
+                     check_existence="at_least_one_exists" version="1"
+                     comment="Check the existence of tcp port defined through services.">
+    <ind:object object_ref="object_var_firewalled_tcp_ports"/>
+  </ind:variable_test>
 
-  <linux:inetlisteningservers_object id="obj_listening_inet_ports" version="2" comment="This object represents a listening services on the system.">
-      <linux:protocol operation="pattern match">^.*$</linux:protocol>
+  <linux:inetlisteningservers_state id="state_firewalled_listening_inet_tcp_ports" version="1"
+                                    comment="Checks listen ports has rule in firewalld">
+    <linux:local_port datatype="int" operation="equals"  var_ref="var_firewalled_tcp_ports"
+                      var_check="at least one"/>
+  </linux:inetlisteningservers_state>
+
+  <linux:inetlisteningservers_state id="state_firewalled_listening_inet_direct_tcp_ports" version="1"
+                                    comment="Checks listen ports has direct port rule in firewalld">
+    <linux:local_port datatype="int" operation="equals"  var_ref="var_firewalled_direct_tcp_ports"
+                      var_check="at least one"/>
+  </linux:inetlisteningservers_state>
+
+  <linux:inetlisteningservers_test id="test_listening_fw_inet_ports_tcp_exist" version="1" check="all"
+                                   check_existence="none_exist"
+                                   comment="Check if any service is listening on tcp ports">
+    <linux:object object_ref="obj_listening_inet_tcp_ports" />
+  </linux:inetlisteningservers_test>
+
+  <linux:inetlisteningservers_object id="obj_listening_inet_tcp_ports" version="2"
```

Comment:
It is possible to include both tcp and udp ports in this object? I believe using this approach you could reduce considerably the OVAL size because less tests would be necessary.

Reply:
good idea :bow: will try to unite them pattern matching protocol

Reply:
This looks a bit more complicated than I thought initially. In general I am trying to find a way to unifiy in one local variable or object variable results from two other objects, i.e. one_object to has the port numbers of firewalled port by service and by number both :
```
  <ind:textfilecontent54_object id="object_firewalled_service" version="1">
    <ind:path>/etc/firewalld/zones/</ind:path>
    <ind:filename operation="pattern match">^.+\.xml$</ind:filename>
    <ind:pattern operation="pattern match">^[\s]*&lt;service name="(\S*)".*$</ind:pattern>
    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
    <filter action="include">state_firewalld_zones_with_interfaces</filter>
  </ind:textfilecontent54_object>

  <ind:textfilecontent54_object id="object_firewalled_direct_tcp_ports" version="1">
    <ind:path>/etc/firewalld/zones/</ind:path>
    <ind:filename operation="pattern match">^.+\.xml$</ind:filename>
    <ind:pattern operation="pattern match">^[\s]*&lt;port\s*(?:(?:protocol="tcp")|)\s*port="(\d+)"\s*(?:(?:protocol="tcp")|)\s*.*$</ind:pattern>
    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
    <filter action="include">state_firewalld_zones_with_interfaces</filter>
  </ind:textfilecontent54_object>

```
, then I can unite udp and tcp, so I can have compared directly udp+tcp listening ports all being checked against common object. Then the whole thing will look much simpler.

Reply:
First thanks for the guideline @marcusburghardt :bow: 
- Reorganized the clauses in the OVAL so object and variables are declared before used - that was very messy I agree 
- also need to make change my emacs mode since it loves indenting xmls not at an given space offset but after the first field in the xml element and therefore got those long lines forcing me to have really lots of lines per definition. 
- Now I think is better, but still the patterns in the regular expression are breaking the line length rule of the style guideline. Didn't find the way to overcome this :(
- Good news is that more thanks to jinja than to oval ;) managed to shorten the definition almost by half defining repeating rules for UDP and TCP. 
- Now I am running fedora automatus on my local machine so hopefully can see the above error also, to be honest could find it in the logs of the continuous integration jobs here